File: LinkedListImplementation/src/DoublyLinkedList.java

import java.time.Period;

public class DoublyLinkedList <T>  {
    public int size = 0;
    private Node<T> head = null;
    private Node<T> tail = null;
    public class Node<T>
    {
        T data;
        Node<T> prev = null;
        Node<T> next = null;
        public Node (T data, Node<T> prev, Node<T> next)
        {
            this.data = data;
            this.prev = prev;
            this.next = next;
        }
        public String toString()
        {
            return data.toString();
        }
    }
    public int getSize()
    {
        return size;
    }
    public boolean isEmpty()
    {
        return size==0;
    }
    public void clear()
    {
        Node<T> trav = head;
        while(trav.next!=null)
        {
            Node<T> travNext = trav.next;
            trav.data = null;
            trav.prev = null;
            trav.next = null;
            trav = travNext;
        }
        head = tail = trav = null;
        size = 0;
    }
    public void add(T elem)
    {
        if(isEmpty())
        {
            addFirst(elem);
        }
        else
        {
            addLast(elem)l;
        }
    }
    public void addFirst(T elem)
    {
        if(isEmpty())
        {
            head = tail = new Node<T>(elem,null,null);
        }
        else
        {
            head.prev = new Node<T>(elem,null,head);
            head = head.prev;
        }
        size++;
    }
    public void addLast(T elem)
    {
        if(isEmpty())
        {
            head = tail = new Node<T>(elem,null,null);
        }
        else
        {
            tail.next = new Node<T>(elem,head,null);
            tail = tail.next;
        }
        size++;
    }
    public T getFirst()
    {
        if(isEmpty()) throw  new RuntimeException("DoublyLinkedList is Empty");
        return head.data;
    }
    public T getLast()
    {
        if(isEmpty()) throw  new RuntimeException("DoublyLinkedList is Empty");
        return tail.data;
    }
    public T removeFirst()
    {
        if(isEmpty()) throw  new RuntimeException("DoublyLinkedList is Empty");
        T thrownData = head.data;
        head = head.next;
        size--;
        if(isEmpty()) tail = null;
        else head.prev = null;
        return thrownData;
    }
    public T removeLast()
    {
        if(isEmpty()) throw  new RuntimeException("DoublyLinkedList is Empty");
        T thrownData = tail.data;
        tail = tail.prev;
        size--;
        if(isEmpty())
        {
            head = null;
        }
        else
        {
            tail.next = null;
        }
        return thrownData;

    }
    private T removeNode(Node<T> node)
    {
        if(node.prev ==null)
        {
            return removeFirst();
        }
        else if (node.next ==null)
        {
            return removeLast();
        }
        else
        {
            node.prev.next = node.next;
            node.next.prev = node.prev;
            T thrownData = node.data;
            node.data = null;
            node.prev = null;
            node.next = null;
            size--;
            return thrownData;
        }
    }
    public T removeAt(int index)
    {
        if(index<0||index>=size)
        {
            throw new RuntimeException("The index shouldn't be zero or be more than size");
        }
        Node<T> trav = head;
        for(int i=0;i<index;i++)
        {
            trav = trav.next;
        }
        return removeNode(trav);
    }
    public T remove(Object obj)
    {
        if(obj==null)
        {
            for(Node<T>trav = head;trav.next!=null;trav=trav.next)
            {
                if(trav.data==null)
                {
                    return removeNode(trav);
                }
            }
        }
        else
        {
            for(Node<T>trav = head;trav.next!=null;trav=trav.next)
            {
                if(obj.equals(trav.data))
                {
                    return removeNode(trav);
                }
            }
        }

        throw new RuntimeException("This object is not detected in linkedlist");
    }
    public int indexOf(Object obj)
    {
        int index=0;
        if(obj==null)
        {
            for(Node<T>trav = head;trav.next!=null;trav=trav.next)
            {
                index++;
                if(trav.data==null)
                {
                    return index;
                }
            }
        }
        else
        {
            for(Node<T>trav = head;trav.next!=null;trav=trav.next)
            {
                index++;
                if(obj.equals(trav.data))
                {
                    return index;
                }
            }
        }

        return -1;
    }
    public String toString()
    {
        String ret = "[ ";
        for(Node<T> trav = head;trav.next!=null;trav=trav.next)
        {
            ret+=String.valueOf(trav.data)+" ";
        }
        ret+=" ]";
        return ret;

    }


}

File: algoOnPythonPractice/DoublyLinkedList.py
class Node:
    def __init__(self,data=None,next=None):
        self.data = None
        self.next = None
        self.prev = None
        if(data!=None):
            self.data = data
    def getData(self):
        return self.data
class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.data = None
        self.tail = None
        self.size=0
    def add(self,value):
        #print(f"Size: {self.size}")
        self.insert(value,self.size)
    def isEmpty(self):
        return self.size==0
    def insert(self,value,index):

        if(index>self.size): raise Exception("index must not larger than size")
        if(self.head == None):
            #print("head")
            self.head = Node(value)
            self.tail = self.head
            self.size+=1
        elif (index ==0):
            self.head.prev = Node(value)
            self.head.prev.next = self.head
            self.head = self.head.prev
        elif (index==self.size):
            #print("tail")
            self.tail.next = Node(value)
            self.tail.next.prev = self.tail
            self.tail = self.tail.next
            self.size+=1
            #print("tail data:" +str(self.tail.data))
        else:
            #print("between")
            cnt = 0
            trav= self.head
            prev = None
            while(trav.next!=None):
                if(cnt == index):
                    break
                prev = trav
                trav = trav.next
                cnt+=1
            prev.next = Node(value)
            prev.next.next = trav
            trav.prev = prev.next
            self.size+=1
    def remove(self,value):
        if(self.isEmpty()):
            raise Exception("Linked List is Empty")
        if not self.search(value):
            raise Exception("Linked is Larger than actual size")
        else:
            trav = self.head
            prev = None
            while(trav.next!=None and trav.data != value):
                prev = trav
                trav=  trav.next
            tmp = trav.data

            prev.next = trav.next
            trav.next.prev = prev
            return tmp
    def search(self,value):
        trav = self.head
        while trav!=None:
            if trav.data == value:
                return True
            trav = trav.next
        return False
    def removeAt(self,index):
        if(self.isEmpty()):
            raise Exception("Linked List is Empty")
        if(index >=self.size):
            raise Exception("Linked is Larger than actual size")
        if(index ==0):
            tmp = self.head.data
            self.head = self.head.next
            self.head.prev = None
            self.size-=1
            return tmp
        elif(index == self.size-1):
            tmp = self.tail.data
            self.tail = self.tail.prev
            self.tail.next = None
            self.size-=1
            return tmp
        else:
            trav = self.head
            prev = None
            cnt=0
            while(trav.next!=None and index != cnt):
                cnt+=1
                prev = trav
                trav=  trav.next
            tmp =trav.data
            prev.next = trav.next
            trav.next.prev = prev
            self.size-=1
            return tmp
    def pop(self):
        return self.removeAt(self.size-1)
    def toString(self):
        ret = ""
        trav = self.head
        while(trav.next!=None):
            #print("current trav val: " + str(trav.data))
            #print("current tail val: " + str(self.tail.data))
            if(trav == self.head):
                ret+=f"[Head: {trav.data}] -> "
            else:
                ret+=f"[{trav.data}] -> "
            trav = trav.next

            #print("Iterated String: "+ ret)
        return ret + f"[Tail: {trav.data}]"
    def display(self):
        print(self.toString())

if __name__ =="__main__":
    dll = DoublyLinkedList()
    dll.add(10)
    dll.add(20)
    dll.add(30)
    dll.add(35)
    dll.display()
    dll.insert(15,0)
    dll.display()
    dll.insert(25,1)
    dll.display()
    dll.pop()
    dll.display()
    dll.removeAt(0)
    dll.display()
    dll.removeAt(2)
    dll.display()



